python/lsst/ctrl/mpexec/cli/cmd/commands.py

```diff
@@ -0,0 +1,103 @@
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import click
+import copy
+
+from lsst.daf.butler.cli.opt import (config_file_option,
+                                     config_option,
+                                     log_level_option)
+from lsst.daf.butler.cli.utils import cli_handle_exception
+from lsst.obs.base.cli.opt import instrument_parameter
+from ..opt import (delete_option,
+                   order_pipeline_option,
+                   pipeline_dot_option,
+                   pipeline_option,
+                   save_pipeline_option,
+                   show_option,
+                   task_option)
+from .. import script
+from ..utils import makePipelineActions
+
+
+instrumentOptionHelp = ("Add an instrument which will be used to load config overrides when defining a "
+                        "pipeline. This must be the fully qualified class name.")
+
+
+class PipetaskCommand(click.Command):
+    def parse_args(self, ctx, args):
+        ctx.obj = copy.copy(args)
+        super().parse_args(ctx, args)
+
+
+@click.command(cls=PipetaskCommand, short_help="Build pipeline definition.")
+@click.pass_context
+@pipeline_option()
+@task_option(multiple=True)
+@delete_option(metavar="LABEL", multiple=True)
+@config_option(metavar="LABEL:NAME=VALUE", multiple=True)
+@config_file_option(help="Configuration override file(s), applies to a task with a given label.",
+                    metavar="LABEL:FILE",
+                    multiple=True)
+@order_pipeline_option()
+@save_pipeline_option()
+@pipeline_dot_option()
+@instrument_parameter(help=instrumentOptionHelp, metavar="instrument", multiple=True)
+@show_option(multiple=True)
+@log_level_option(defaultValue=None)
+def build(ctx, *args, **kwargs):
+    """Build and optionally save pipeline definition.
+
+    This does not require input data to be specified.
+    """
+    # The pipeline actions (task, delete, config, config_file, and instrument)
+    # must be handled in the order they appear on the command line, but the CLI
+    # specification gives them all different option names. So, instead of using
+    # the individual action options as they appear in kwargs (because
+    # invocation order can't be known), we capture the CLI arguments by
+    # overriding `click.Command.parse_args` and save them in the Context's
+    # `obj` parameter. We use `makePipelineActions` to create a list of
+    # pipeline actions from the CLI arguments and pass that list to the script
+    # function using the `pipeline_actions` kwarg name, and remove the action
+    # options from kwargs.
+    for pipelineAction in (task_option.optionKey, delete_option.optionKey, config_option.optionKey,
+                           config_file_option.optionKey, instrument_parameter.optionKey):
+        kwargs.pop(pipelineAction)
+    kwargs['pipeline_actions'] = makePipelineActions(ctx.obj)
```

Comment:
This looks a bit hackish. Could you instead just define a callback for each option that updates a common list? Or maybe click already has a way to append values from multiple options to the same parameter? Then you could use the same approach as I did in CmdLineParser - use those `_ACTION...` as type converters for individual options?

Reply:
The click api really doesn't support any of these approaches. I think the best way would be to pivot to chaining subcommands, and making each of the action options a type of subcommand. Discussing that in Slack now.